### ISSUE TITLE
Implement Codable template

### ISSUE THREAD

**Issue by ilyapuchka** [krzysztofzablocki/Sourcery#353]
With introduction of native encoding/decoding with Swift 4 it might seem for some users that Sourcery is loosing its relevance in this domain. Though it is true for simplest cases for more complex cases requiring keys renaming and manual encoding/decoding it is still very relevant. This template should show how Sourcery still can be used to eliminate boilerplate around that. 

- [ ] Inline Coding Keys so that user only defines keys which do not match
- [ ] Support different key decoding strategies, not based on standard library strategy, but instead used by Sourcery to generate values for keys (compile-time vs run-time)

```swift
enum CodingKeys: String, CodingKey, UppercaseFirstCodingKeyStrategy {
  case id = "identifier"
  // sourcery:auto:inline: Type.CodingKeys
  the rest of keys goes here with raw values computed using decoding strategy
  // sourcery:auto:inline:end
}
```
- [ ] Generate constructors in extensions (inlined or in separate file), so that structs still have member-wise initialisers
- [ ] ~~Generated constructor should call member-wise constructor (generate one private if needed, if there is any other constructor that does not match)~~

```swift
struct Type {
   ... //properties
   enum CodingKeys { ... }
}

// sourcery:inline: Type.Decoding
extension Type: Swift.Decoding { // add conformance if not conforming yet
  init(from decoder: Swift.Decoder) throws { ... }
}
// sourcery:end
```

- [ ] Support custom coding methods in form of `decode<property name>`/`encode<property name>` instead of using annotations
- [ ] helper methods for safely decoding arrays, formatted dates etc.

**Comment by yoavschwartz** [comment]
Came here looking for this :) I have been working with swift 4 for a while now, and for 95% of the objects I have to write `CodingKeys` by hand to override one or 2 of the keys, and about 50% I need to write `initFromDecoder` by hand since some verification/processing of data is needed. I don't think `Codable` will kill Sourcery's usability for JSON parsing, I think it will probably be 10 times more useful! Can't wait for this feature, I really enjoy using Sourcery. I tried to implement this myself, but I ran into a problem. I wrote a basic stencil that creates an `initFrom(decoder:)` in the most naive way, but then I need to write some code to customise the method itself after it is generated. How would I go about mixing Sourcery generated code with hand written code?

**Comment by rodrigoruiz** [comment]
Any updates on this? Especially `Codable` for `enum`s.

The [Codable Enums](https://littlebitesofcocoa.com/318-codable-enums) from documentation doesn't show how to use Sourcery to implement it as it claims, it only shows how to manually make an `enum` `Codable`.

**Comment by RomanTysiachnik** [comment]
Well you can't make enum Codable by default right now. And yeah, it would've been great not to write decode and encode for each enum, especially when you have a lot of them in hierarchy. Or at least to generate them.

**Comment by thevest** [comment]
Just implemented a basic version for my own project, it doesn't handle a lot of edge cases / is pretty rough around the edges, nonetheless for anyone that may be interested...

Template

```
import Foundation

// https://github.com/SwiftGen/StencilSwiftKit/blob/master/Documentation/tag-set.md
// https://github.com/kylef/Stencil
// https://cdn.rawgit.com/krzysztofzablocki/Sourcery/master/docs/index.html

{% macro modelName name %}{{ name | replace:"Mappable","Model" }}{% endmacro %}

{% macro immutableVariables variables %}
  {% for var in variables %}
  let {{ var.name }}: {{ var.typeName }}
  {% endfor %}
{% endmacro %}

{% macro realmConvert model %}
  {% set rmodel %}{{ model|replace:"Model","" }}{% endset %}
  public override func asRealm() -> {{ rmodel }} {
    return {{ rmodel }}(value: asDictionary())
  }
{% endmacro %}

{% macro codingKeys variables %}
  private enum CodingKeys: String, CodingKey {
    {% for var in variables %}
    case {{ var.name }} {% if var.annotations.coding %} = "{{ var.annotations.coding }}" {% endif %}
    {% endfor %}
  }
{% endmacro %}

{% macro decodeVar var %}
  {% set itype %}{{ var.annotations.incomingType }}{% endset %}
  {% if var.typeName.name == "Bool" and itype == "Int" %} {# its a 0 or 1 #}
    {{ var.name }} = try container.decodeIfPresent(Int.self, forKey: .{{ var.name }}) == 1
  {% elif var.typeName.name == "Date" and itype == "String" %} {# iso date string #}
    if
      let {{ var.name }} = try container.decodeIfPresent(String.self, forKey: .{{ var.name }}),
      let date = {{ var.name }}.dateFromISO8601 {
      self.{{ var.name }} = date
    } else {
      self.{{ var.name }} = Date()
    }
  {% endif %}
{% endmacro %}

{# cant get spacing right without doing it like this, better way when possible #}
{% macro decodeWithDefault var %}
  {% if var.type.name == "String" %}
    {{ var.name }} = try container.decodeIfPresent({{ var.typeName }}.self, .{{ var.name }}) ?? ""
  {% elif var.type.name == "Bool" %}
    {{ var.name }} = try container.decodeIfPresent({{ var.typeName }}.self, .{{ var.name }}) ?? false
  {% elif var.type.name == "Int" %}
    {{ var.name }} = try container.decodeIfPresent({{ var.typeName }}.self, .{{ var.name }}) ?? 0
  {% elif var.type.name == "Date" %}
    {{ var.name }} = try container.decodeIfPresent({{ var.typeName }}.self, .{{ var.name }}) ?? Date()
  {% endif %}
{% endmacro %}

{% macro kvoSubWithDefault var %}
  {% if var.type.name == "String" %}
    {{ var.name }} = object["{{ var.name }}"] as? {{ var.typeName.name }} ?? ""
  {% elif var.type.name == "Bool" %}
    {{ var.name }} = object["{{ var.name }}"] as? {{ var.typeName.name }} ?? false
  {% elif var.type.name == "Int" or var.type.name == "Double" or var.type.name == "Float" %}
    {{ var.name }} = object["{{ var.name }}"] as? {{ var.typeName.name }} ?? 0
  {% elif var.type.name == "Date" %}
    {{ var.name }} = object["{{ var.name }}"] as? {{ var.typeName.name }} ?? Date()
  {% endif %}
{% endmacro %}

{% macro decodableInitializer variables %}
  public required init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    {% for var in variables where var|annotated:"incomingType" %}
    {% call decodeVar var %}
    {% endfor %}
    {% for var in variables where var|!annotated:"incomingType" %}
    {% call decodeWithDefault var %}
    {% endfor %}
    try super.init(from: decoder)
  }
{% endmacro %}

{% macro convertVarForObject var %}
  {% set itype %}{{ var.annotations.incomingType }}{% endset %}
  {% if var.typeName.name == "Bool" and itype == "Int" %} {# its a 0 or 1 #}
    {{ var.name }} = (object["{{ var.name }}"] as? Int) == 1
  {% elif var.typeName.name == "Date" and itype == "String" %} {# iso date string #}
    if let {{ var.name }} = object["{{ var.name }}"] as? NSNumber {
      self.{{ var.name }} = Date.fromTimestamp(Timestamp({{ var.name }}.int64Value))
    } else {
      self.{{ var.name }} = Date()
    }
  {% endif %}
{% endmacro %}

{% macro objectInitializer variables %}
  public required init(with object: INRBaseObject) {
    {% for var in variables where var|annotated:"incomingType" %}
    {% call convertVarForObject var %}
    {% endfor %}
    {% for var in variables where var|!annotated:"incomingType" %}
    {% call kvoSubWithDefault var %}
    {% endfor %}
    super.init(with: object)
  }
{% endmacro %}

{% macro convertVarForEncoder var %}
  {% set itype %}{{ var.annotations.incomingType }}{% endset %}
  {% if var.typeName.name == "Bool" and itype == "Int" %} {# its a 0 or 1 #}
    try container.encode({{ var.name }} ? 1 : 0, forKey: .{{ var.name }})
  {% elif var.typeName.name == "Date" and itype == "String" %} {# iso date string #}
    try container.encode({{ var.name }}.iso8601, forKey: .{{ var.name }})
  {% endif %}
{% endmacro %}

{% macro encode variables %}
  public override func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    {% for var in variables where var|annotated:"incomingType" %}
    {% call convertVarForEncoder var %}
    {% endfor %}
    {% for var in variables where var|!annotated:"incomingType" %}
    try container.encode({{ var.name }}, forKey: .{{ var.name }})
    {% endfor %}
    try super.encode(to: encoder)
  }
{% endmacro %}

{% macro asDictionary variables %}
  public override func asDictionary() -> [String: Any] {
    var dictionary = super.asDictionary()
    {% for var in variables %}
    dictionary["{{ var.name }}"] = {{ var.name }}
    {% endfor %}
    return dictionary
  }
{% endmacro %}

{% for type in types.based.ModelMappable where type|protocol and not type.name == "ModelMappable" %}
{% set model %}{% call modelName type.name %}{% endset %}

// MARK: {{ model }} Mappable
public final class {{ model }}: BaseModel, {{ type.name }} {
  {% call immutableVariables type.allVariables %}

  // MARK: {{ model }} CodingKeys
  {% call codingKeys type.allVariables %}

  // MARK: {{ model }} Decodable
  {% call decodableInitializer type.allVariables %}

  // MARK: {{ model }} Object Init
  {% call objectInitializer type.allVariables %}

  // MARK: {{ model }} Encodable
  {% call encode type.allVariables %}

  // MARK: {{ model }} Dictionary
  {% call asDictionary type.allVariables %}

  // MARK: {{ model }} Realm Convert
  {% call realmConvert model %}
}

{% endfor %}
```


```
public protocol ModelMappable { }

public protocol ExampleMappable: ModelMappable {
  // sourcery: coding="some_variable_named_something_else"
  var someVariable: String { get set }
  // sourcery: coding="some_variable_id", incomingType="Int"
  var someBoolThatsAnIntFromTheServer: Bool { get set }
}
```

Produces

```
// MARK: ExampleModel Mappable
public final class ExampleModel: BaseModel, ExampleMappable {
  let someVariable: String
  let someBoolThatsAnIntFromTheServer: Bool

  // MARK: ExampleModel CodingKeys
  private enum CodingKeys: String, CodingKey {
    case someVariable  = "some_variable_named_something_else" 
    case someBoolThatsAnIntFromTheServer  = "some_variable_id" 
  }

  // MARK: ExampleModel Decodable
  public required init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    someBoolThatsAnIntFromTheServer = try container.decodeIfPresent(Int.self, forKey: .someBoolThatsAnIntFromTheServer) == 1
    someVariable = try container.decodeIfPresent(String.self, .someVariable) ?? ""
    try super.init(from: decoder)
  }

  // MARK: ExampleModel Object Init
  public required init(with object: INRBaseObject) {
    someBoolThatsAnIntFromTheServer = (object["someBoolThatsAnIntFromTheServer"] as? Int) == 1
    someVariable = object["someVariable"] as? String ?? ""
    super.init(with: object)
  }

  // MARK: ExampleModel Encodable
  public override func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encode(someBoolThatsAnIntFromTheServer ? 1 : 0, forKey: .someBoolThatsAnIntFromTheServer)
    try container.encode(someVariable, forKey: .someVariable)
    try super.encode(to: encoder)
  }

  // MARK: ExampleModel Dictionary
  public override func asDictionary() -> [String: Any] {
    var dictionary = super.asDictionary()
    dictionary["someVariable"] = someVariable
    dictionary["someBoolThatsAnIntFromTheServer"] = someBoolThatsAnIntFromTheServer
    return dictionary
  }

  // MARK: ExampleModel Realm Convert
  public override func asRealm() -> Example {
    return Example(value: asDictionary())
  }
}
```

I just found this project yesterday and it is really awesome by the way! @krzysztofzablocki What do you think about moving this repo to an organization and then creating a sort of community templates project to accompany it?

**Comment by krzysztofzablocki** [comment]
@thevest thanks for sharing the template, we discussed organization before but we don't want to manage multiple projects as its PITA, we already share common startup templates in this repo and have tests for them that run each time Sourcery is changed which helps with maintaining them

@thevest would you mind adding it to our bundled templates along with some basic tests? even if it doesn't handle all cases it would be good start for anyone needing Codable support :)

**Comment by ilyapuchka** [comment]
Resolved with #599
